### PR TITLE
[POAE7-2535] Support Case/IF Function for Arrow Format with Null value

### PR DESCRIPTION
**Pull request by haiweizh**
### What changes were proposed in this pull request?

Support Case/IF Function for Arrow Format with Null value.

### Why are the changes needed?

Support Case/IF Function for Arrow Format with Null value.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Uts.

### Which label does this PR belong to?

cider,test

### CODE REVIEW

**Comment by haiweizh** [comment]
Support Case/IF Function for Arrow Format with Null value. Plz help to take a review. @yma11 @jikunshang @spevenhe

File: cider/tests/functionality/CiderCaseWhenTest.cpp [hunk 1]
```diff
@@ -89,7 +89,7 @@ class CiderArrowCaseWhenSequenceTestBase : public CiderTestBase {
   CiderArrowCaseWhenSequenceTestBase() {
     table_name_ = "test";
     create_ddl_ =
-        R"(CREATE TABLE test(col_int INTEGER NOT NULL, col_bigint BIGINT NOT NULL, col_double DOUBLE NOT NULL, col_float FLOAT NOT NULL);)";
+        R"(CREATE TABLE test(col_int INTEGER, col_bigint BIGINT, col_double DOUBLE, col_float FLOAT);)";
```
Comment (yma11) [inline cider/tests/functionality/CiderCaseWhenTest.cpp:92]: Seems these `NOT NULL` shouldn't be deleted.

File: cider/exec/template/CaseIR.cpp [hunk 2]
```diff
@@ -178,18 +178,23 @@ std::unique_ptr<CodegenColValues> CodeGenerator::codegenCaseExpr(
...
-    }
+    CHECK(if_expr);
     const auto when_lv = toBool(if_expr->getValue());
+    const auto when_null_lv = if_expr->getNull();
+    CHECK(when_lv);
+    CHECK(when_null_lv);
+    llvm::Value* filter_lv = cgen_state_->llBool(true);
+    filter_lv = cgen_state_->ir_builder_.CreateAnd(
+        filter_lv, cgen_state_->ir_builder_.CreateNot(when_null_lv));
+    auto final_when_lv = cgen_state_->ir_builder_.CreateAnd(filter_lv, when_lv);
```
Comment (spevenhe) [inline cider/exec/template/CaseIR.cpp:197]: Could you paste the generated IR of this part?

Comment (haiweizh) [reply cider/exec/template/CaseIR.cpp:197]: %0 = call i8* @cider_ColDecoder_extractArrowBuffersAt(i8* %col_buf0, i64 1)
  %1 = call i8* @cider_ColDecoder_extractArrowBuffersAt(i8* %col_buf0, i64 0)
  %2 = call i64 @fixed_width_int_decode(i8* %0, i32 4, i64 %pos)
  %3 = trunc i64 %2 to i32
  %4 = call i1 @check_bit_vector_clear(i8* %1, i64 %pos)
  %5 = icmp slt i32 %3, %arg_literal_0
  %6 = xor i1 %4, true
  %7 = and i1 true, %6
  %8 = and i1 %7, %5
  br i1 %8, label %then, label %else

Comment (haiweizh) [reply cider/exec/template/CaseIR.cpp:197]: Like this, do the filter.

File: cider/tests/functionality/CiderCaseWhenTest.cpp (hunk 1 above)
Comment (haiweizh) [reply cider/tests/functionality/CiderCaseWhenTest.cpp:92]: Xinyi plz help to explain. @spevenhe

File: cider/exec/template/CaseIR.cpp [hunk 3]
```diff
@@ -198,28 +203,34 @@ std::unique_ptr<CodegenColValues> CodeGenerator::codegenCaseExpr(
...
+    CHECK(then_expr);
     auto then_bb_lvs = then_expr->getValue();
+    auto then_bb_null_lvs = then_expr->getNull();
     if (is_real_str) {
       // FIXME(haiwei): [POAE7-2457] ,blocking by [POAE7-2415]
       CIDER_THROW(CiderCompileException,
                   "String type case when is not currently supported.");
     } else {
       then_lvs.push_back(then_bb_lvs);
+      then_null_lvs.push_back(then_bb_null_lvs);
```
Comment (yma11) [inline cider/exec/template/CaseIR.cpp:215]: Can you add some comments to address how the null_lvs determined in CaseIR?

File: cider/tests/functionality/CiderCaseWhenTest.cpp (hunk 1 above)
Comment (haiweizh) [reply cider/tests/functionality/CiderCaseWhenTest.cpp]: Done. Restore NOT NULL.

File: cider/exec/template/CaseIR.cpp (hunk 3 above)
Comment (haiweizh) [reply cider/exec/template/CaseIR.cpp]: Rewrite these code for NOT NULL columns.

**Review by yma11 — APPROVED** [review]

**Review by spevenhe — APPROVED** [review]
LGTM

**Pull request merged**